Upstream-Status: Backport

This is a backport from Chromium's FFMPEG git repository.

Signed-off-by: Raphael Kubo da Costa <raphael.kubo.da.costa@intel.com>
---
From 901b874d864bb5ee7b4758812a1a3665d53435e3 Mon Sep 17 00:00:00 2001
From: Arun Patole <apatole@nvidia.com>
Date: Mon, 29 Aug 2016 10:41:50 +0530
Subject: [PATCH] Fix Chrome ARM64 linux compilation errors

For GYP build,
We get HAVE_VFP_ARGS redefinition error.
To fix it, just disabling HAVE_VFP_ARGS definition in
chromium/config/Chromium/linux/arm64/config.h

For GN Build,
We get name conflict for videodsp.o as it is generated twice in same
folder using libavcodec/videodsp.c and libavcodec/aarch64/videodsp.S.
This change renames videodsp.S to videodsp_aarch64.S as a workaround
to this build issue.

BUG=613452

Change-Id: Iccb811ee6e3d10f888a9d1e5efb65055b409d23c
Reviewed-on: https://chromium-review.googlesource.com/360036
Reviewed-by: Chrome Cunningham <chcunningham@chromium.org>
---
 chromium/config/Chrome/linux/arm64/config.h   |  2 +-
 chromium/config/Chromium/linux/arm64/config.h |  2 +-
 chromium/patches/README                       |  6 ++++++
 chromium/scripts/build_ffmpeg.py              |  2 +-
 ffmpeg_generated.gni                          |  2 +-
 libavcodec/aarch64/videodsp.S                 | 28 ---------------------------
 libavcodec/aarch64/videodsp_aarch64.S         | 28 +++++++++++++++++++++++++++
 7 files changed, 38 insertions(+), 32 deletions(-)
 delete mode 100644 libavcodec/aarch64/videodsp.S
 create mode 100644 libavcodec/aarch64/videodsp_aarch64.S

diff --git a/third_party/ffmpeg/chromium/config/Chrome/linux/arm64/config.h b/third_party/ffmpeg/chromium/config/Chrome/linux/arm64/config.h
index f9a30d5..ea27021 100644
--- a/third_party/ffmpeg/chromium/config/Chrome/linux/arm64/config.h
+++ b/third_party/ffmpeg/chromium/config/Chrome/linux/arm64/config.h
@@ -344,7 +344,7 @@
 #define HAVE_RSYNC_CONTIMEOUT 1
 #define HAVE_SYMVER_ASM_LABEL 0
 #define HAVE_SYMVER_GNU_ASM 1
-#define HAVE_VFP_ARGS 0
+/* #define HAVE_VFP_ARGS 0 -- Disabled to allow softfp/hardfp selection at gyp time */
 #define HAVE_XFORM_ASM 0
 #define HAVE_XMM_CLOBBERS 0
 #define HAVE_CONDITION_VARIABLE_PTR 0
diff --git a/third_party/ffmpeg/chromium/config/Chromium/linux/arm64/config.h b/third_party/ffmpeg/chromium/config/Chromium/linux/arm64/config.h
index 478c8cf..a4ef4f4 100644
--- a/third_party/ffmpeg/chromium/config/Chromium/linux/arm64/config.h
+++ b/third_party/ffmpeg/chromium/config/Chromium/linux/arm64/config.h
@@ -344,7 +344,7 @@
 #define HAVE_RSYNC_CONTIMEOUT 1
 #define HAVE_SYMVER_ASM_LABEL 0
 #define HAVE_SYMVER_GNU_ASM 1
-#define HAVE_VFP_ARGS 0
+/* #define HAVE_VFP_ARGS 0 -- Disabled to allow softfp/hardfp selection at gyp time */
 #define HAVE_XFORM_ASM 0
 #define HAVE_XMM_CLOBBERS 0
 #define HAVE_CONDITION_VARIABLE_PTR 0
diff --git a/third_party/ffmpeg/chromium/patches/README b/third_party/ffmpeg/chromium/patches/README
index 898eb75..8f758eb 100644
--- a/third_party/ffmpeg/chromium/patches/README
+++ b/third_party/ffmpeg/chromium/patches/README
@@ -97,3 +97,9 @@ Current patches:
   Changes to libavformat/riffdec.c
     Work around VC++ 2015 Update 1 compiler bug, crbug.com/581768 and
     https://connect.microsoft.com/VisualStudio/feedback/details/2291638
+
+  Changes to libavcodec/aarch64/videodsp.S
+    Renamed videodsp.S to videodsp_aarch64.S. This is to fix GN build errors
+    for arm64 linux. There is a conflict for generated videodsp.o as both
+    libavcodec/videodsp.c and libavcodec/aarch64/videodsp.S tries to generate
+    videodsp.o at same location.
diff --git a/third_party/ffmpeg/chromium/scripts/build_ffmpeg.py b/third_party/ffmpeg/chromium/scripts/build_ffmpeg.py
index 2e0d342..881494a 100755
--- a/third_party/ffmpeg/chromium/scripts/build_ffmpeg.py
+++ b/third_party/ffmpeg/chromium/scripts/build_ffmpeg.py
@@ -232,7 +232,7 @@ def BuildFFmpeg(target_os, target_arch, host_os, host_arch, parallel_jobs,
           'Host arch : %s\n'
           'Target arch : %s\n' % (host_os, target_os, host_arch, target_arch))
 
-  if target_arch in ('arm', 'arm-neon'):
+  if target_arch in ('arm', 'arm-neon', 'arm64'):
     RewriteFile(
         os.path.join(config_dir, 'config.h'),
         r'(#define HAVE_VFP_ARGS [01])',
diff --git a/third_party/ffmpeg/ffmpeg_generated.gni b/third_party/ffmpeg/ffmpeg_generated.gni
index e952679..42c0bbf 100644
--- a/third_party/ffmpeg/ffmpeg_generated.gni
+++ b/third_party/ffmpeg/ffmpeg_generated.gni
@@ -459,7 +459,7 @@ if ((is_linux && current_cpu == "arm64" && ffmpeg_branding == "Chrome") || (is_l
   ffmpeg_gas_sources += [
     "libavcodec/aarch64/autorename_libavcodec_aarch64_h264pred_neon.S",
     "libavcodec/aarch64/autorename_libavcodec_aarch64_hpeldsp_neon.S",
-    "libavcodec/aarch64/videodsp.S",
+    "libavcodec/aarch64/videodsp_aarch64.S",
   ]
 }
 
diff --git a/third_party/ffmpeg/libavcodec/aarch64/videodsp.S b/libavcodec/aarch64/videodsp.S
deleted file mode 100644
index 24067cc..0000000
--- a/third_party/ffmpeg/libavcodec/aarch64/videodsp.S
+++ /dev/null
@@ -1,28 +0,0 @@
-/*
- * This file is part of FFmpeg.
- *
- * FFmpeg is free software; you can redistribute it and/or
- * modify it under the terms of the GNU Lesser General Public
- * License as published by the Free Software Foundation; either
- * version 2.1 of the License, or (at your option) any later version.
- *
- * FFmpeg is distributed in the hope that it will be useful,
- * but WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
- * Lesser General Public License for more details.
- *
- * You should have received a copy of the GNU Lesser General Public
- * License along with FFmpeg; if not, write to the Free Software
- * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
- */
-
-#include "libavutil/aarch64/asm.S"
-
-function ff_prefetch_aarch64, export=1
-        subs            w2,  w2,  #2
-        prfm            pldl1strm, [x0]
-        prfm            pldl1strm, [x0,  x1]
-        add             x0,  x0,  x1,  lsl #1
-        b.gt            X(ff_prefetch_aarch64)
-        ret
-endfunc
diff --git a/third_party/ffmpeg/libavcodec/aarch64/videodsp_aarch64.S b/third_party/ffmpeg/libavcodec/aarch64/videodsp_aarch64.S
new file mode 100644
index 0000000..24067cc
--- /dev/null
+++ b/third_party/ffmpeg/libavcodec/aarch64/videodsp_aarch64.S
@@ -0,0 +1,28 @@
+/*
+ * This file is part of FFmpeg.
+ *
+ * FFmpeg is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * FFmpeg is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with FFmpeg; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include "libavutil/aarch64/asm.S"
+
+function ff_prefetch_aarch64, export=1
+        subs            w2,  w2,  #2
+        prfm            pldl1strm, [x0]
+        prfm            pldl1strm, [x0,  x1]
+        add             x0,  x0,  x1,  lsl #1
+        b.gt            X(ff_prefetch_aarch64)
+        ret
+endfunc
-- 
2.7.4

